packages/autocomplete-core/src/utils/createCancelablePromise.ts

```diff
@@ -0,0 +1,161 @@
+  resolve: (value: TValue | PromiseLike<TValue>) => void,
+  reject: (reason?: any) => void
+) => void;
+
+type CreateInternalCancelablePromiseParams<TValue> = {
+  executor?: PromiseExecutor<TValue>;
+  promise?: Promise<TValue>;
+  initialState?: CancelablePromiseState;
+};
+
+function createInternalCancelablePromise<TValue>({
+  executor = noop,
+  initialState = createInitialState(),
+  promise = new Promise<TValue>((resolve, reject) => {
+    return executor(resolve, reject);
+  }),
+}: CreateInternalCancelablePromiseParams<TValue>): CancelablePromise<TValue> {
+  const state = initialState;
+
+  return {
+    then(onfulfilled, onrejected) {
+      return createCancelable(
+        promise.then(
+          createCallback(onfulfilled, state, promise),
+          createCallback(onrejected, state, promise)
+        ),
+        state
+      );
+    },
+    catch(onrejected) {
+      return createCancelable(
+        promise.catch(createCallback(onrejected, state, promise)),
+        state
+      );
+    },
+    finally(onfinally, runWhenCanceled) {
+      if (runWhenCanceled && onfinally) {
+        state.onCancelList.push(onfinally);
+      }
+
+      return createCancelable<TValue>(
+        promise.finally(
+          createCallback(
+            onfinally &&
+              (() => {
+                if (runWhenCanceled) {
+                  state.onCancelList = [];
+                }
+
+                return onfinally();
+              }),
+            state,
+            promise
+          )
+        ),
+        state
+      );
+    },
+    cancel() {
+      state.isCanceled = true;
+
+      state.onCancelList.forEach((callback) => {
+        callback();
+      });
+    },
+    isCanceled() {
+      return state.isCanceled === true;
+    },
+  };
+}
+
+export type CancelablePromise<TValue> = {
+  then<TResultFulfilled = TValue, TResultRejected = never>(
+    onfulfilled?:
+      | ((
+          value: TValue
+        ) =>
+          | TResultFulfilled
+          | PromiseLike<TResultFulfilled>
+          | CancelablePromise<TResultFulfilled>)
+      | undefined
+      | null,
+    onrejected?:
+      | ((
+          reason: any
+        ) =>
+          | TResultRejected
+          | PromiseLike<TResultRejected>
+          | CancelablePromise<TResultRejected>)
+      | undefined
+      | null
+  ): CancelablePromise<TResultFulfilled | TResultRejected>;
+  catch<TResult = never>(
+    onrejected?:
+      | ((
+          reason: any
+        ) => TResult | PromiseLike<TResult> | CancelablePromise<TResult>)
+      | undefined
+      | null
+  ): CancelablePromise<TValue | TResult>;
+  finally(
+    onfinally?: (() => void) | undefined | null,
+    runWhenCanceled?: boolean
+  ): CancelablePromise<TValue>;
+  cancel(): void;
+  isCanceled(): boolean;
+};
+
+export function createCancelablePromise<TValue>(
+  executor: PromiseExecutor<TValue>
+): CancelablePromise<TValue> {
+  return createInternalCancelablePromise({ executor });
```

Comment:
Why do we need this indirection?

Reply:
It allows us to pass an `initialState` (see `createCancelable`). This is only usefull internally, we don't need this API in consuming code.